src/core/qgsdiagramrenderer.cpp

```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
+QgsStackedDiagramRenderer *QgsStackedDiagramRenderer::clone() const
+{
+  return new QgsStackedDiagramRenderer( *this );
+}
+
+QSizeF QgsStackedDiagramRenderer::sizeMapUnits( const QgsFeature &feature, const QgsRenderContext &c ) const
+{
+  QSizeF stackedSize( 0, 0 );
+  int enabledDiagramCount = 0;  // We'll add spacing only for enabled subDiagrams
+
+  // Iterate renderers. For each renderer, get the diagram
+  // size for the feature and add it to the total size
+  // accounting for stacked diagram defined spacing
+  for ( const auto &subRenderer : std::as_const( mDiagramRenderers ) )
+  {
+    QSizeF size = subRenderer->sizeMapUnits( feature, c );
+
+    if ( size.isValid() )
+    {
+      enabledDiagramCount++;
+      switch ( mSettings.stackedDiagramMode )
+      {
+        case QgsDiagramSettings::Horizontal:
+          stackedSize.setWidth( stackedSize.width() + size.width() );
+          stackedSize.setHeight( std::max( stackedSize.height(), size.height() ) );
+          break;
+
+        case QgsDiagramSettings::Vertical:
+          stackedSize.setWidth( std::max( stackedSize.width(), size.width() ) );
+          stackedSize.setHeight( stackedSize.height() + size.height() );
+          break;
+      }
+    }
+  }
+
+  if ( stackedSize.isValid() )
+  {
+    const double spacing = c.convertToMapUnits( mSettings.stackedDiagramSpacing(), mSettings.stackedDiagramSpacingUnit(), mSettings.stackedDiagramSpacingMapUnitScale() );
+
+    switch ( mSettings.stackedDiagramMode )
+    {
+      case QgsDiagramSettings::Horizontal:
+        stackedSize.scale( stackedSize.width() + spacing * ( enabledDiagramCount - 1 ), stackedSize.height(), Qt::IgnoreAspectRatio );
+        break;
+
+      case QgsDiagramSettings::Vertical:
+        stackedSize.scale( stackedSize.width(), stackedSize.height() + spacing * ( enabledDiagramCount - 1 ), Qt::IgnoreAspectRatio );
+        break;
+    }
+  }
+  return stackedSize;
+}
+
+void QgsStackedDiagramRenderer::renderDiagram( const QgsFeature &feature, QgsRenderContext &c, QPointF pos, const QgsPropertyCollection &properties ) const
+{
+  if ( !mDiagram )
+  {
+    return;
+  }
+
+  QPointF newPos = pos; // Each subdiagram will have its own newPos
+
+  // Get subrenderers sorted by mode (vertical diagrams are returned backwards)
+  const QList< QgsDiagramRenderer * > stackedRenderers = renderers( true );
+
+  for ( const auto &stackedRenderer : stackedRenderers )
+  {
+    if ( stackedRenderer->rendererName() == QStringLiteral( "Stacked" ) )
+    {
+      // Nested stacked diagrams will use this recursion
+      stackedRenderer->renderDiagram( feature, c, newPos, properties );
+      continue;
+    }
+
+    QgsDiagramSettings s;
+    if ( !stackedRenderer->diagramSettings( feature, c, s ) )
+    {
+      continue;
+    }
+
+    if ( !s.enabled )
+    {
+      continue;
+    }
+
+    if ( s.scaleBasedVisibility )
+    {
+      double maxScale = s.maximumScale;
+      if ( maxScale > 0 && c.rendererScale() < maxScale )
```

Comment:
isn't one the boundaries inclusive for scale visibility?

Reply:
Good catch! I think you've found a good source of unexpected behaviors. I had copied this block from the current `QgsVectorLayerDiagramProvider::registerDiagram()`.

After some GUI tests with a given scale (say 1:50.000), the scale is sometimes a round number and sometimes a number like `49999.999999998996` or `50000.000000003354` (I could reproduce it when maximizing/restoring the QGIS main window), so those checks were failing from time to time, which could make the diagrams disappear when they're expected to be rendered and vice versa. 

I added checks using qgsDoubleNear() to take non-round values into account.
Applied in https://github.com/qgis/QGIS/pull/58568/commits/b8f976012ab383e29e7156ff75ea3ea836b99e06